File: gui/DesignerV1_55/ToggleDemo/ToggleDemo.pas
Program ToggleDemo;

{*****************************************************}
{*                                                   *}
{*   This demo shows how to use Boolean type gadgets *}
{*  to create mutually exclusive gadgets.            *}
{*   When a Gadget is pressed it must de-select the  *}
{*  other Gadgets and select itself.                 *}
{*                                                   *}
{*****************************************************}

Uses exec,intuition,gadtools,graphics,amiga,diskfont,
     workbench,utility,toggledemowin;

const
  FirstGad  : string = 'First Option'#0;
  SecondGad : string = 'Second Option'#0;
  ThirdGad  : string = 'Third Option'#0;
var
  done      : boolean;
  class     : long;
  code      : word;
  pimsg     : pintuimessage;
  dummy     : long;
  pgsel     : pgadget;
  gadnumber : word;
begin
  if openlibs then
    begin
      if makeimages then
        begin
          if openwindowmainwindow then
            begin
              done:=false;
              repeat
                dummy:=wait(bitmask(mainwindow^.userport^.mp_sigbit));
                pimsg:=gt_getimsg(mainwindow^.userport);
                while(pimsg<>nil) do
                  begin
                    class:=pimsg^.class;
                    code:=pimsg^.code;
                    pgsel:=pgadget(pimsg^.iaddress);  { do not reference unless gadgetup or gadgetdown }
                    gt_replyimsg(pimsg);
                    gadnumber:=99;
                    case class of
                      idcmp_closewindow : 
                        done:=true;
                      idcmp_vanillakey :
                        case upcase(chr(code)) of
                          'F' : gadnumber:=FirstGadget;
                          'S' : gadnumber:=SecondGadget;
                          'T' : gadnumber:=ThirdGadget;
                         end;
                      idcmp_gadgetdown :
                        gadnumber:=pgsel^.gadgetid;
                     end;
                    case gadnumber of
                      FirstGadget :
                        begin
                          {* Remove gadgets from window *}
                          dummy:=RemoveGList(mainwindow,mainwindowglist,~0);
                          {* Change Gadget Flags *}
                          mainwindowgads[FirstGadget]^.Flags:=mainwindowgads[FirstGadget]^.Flags or GFLG_Selected;
                          mainwindowgads[SecondGadget]^.Flags:=mainwindowgads[SecondGadget]^.Flags and ~GFLG_Selected;
                          mainwindowgads[ThirdGadget]^.Flags:=mainwindowgads[ThirdGadget]^.Flags and ~GFLG_Selected;
                          {* Put Gadgets Back in window *}
                          dummy:=AddGList(mainwindow,mainwindowglist,dummy,~0,nil);
                          {* Refresh Gadgets *}
                          RefreshGList(MainWindowGList,MainWindow,nil,~0);
                          gt_setsinglegadgetattr(mainwindowgads[DisplayGadget],mainwindow,
                                                 GTTX_Text,long(@FirstGad[1]));
                        end;
                      SecondGadget :
                        begin
                          dummy:=RemoveGList(mainwindow,mainwindowglist,~0);
                          mainwindowgads[FirstGadget]^.Flags:=mainwindowgads[FirstGadget]^.Flags and ~GFLG_Selected;
                          mainwindowgads[SecondGadget]^.Flags:=mainwindowgads[SecondGadget]^.Flags or GFLG_Selected;
                          mainwindowgads[ThirdGadget]^.Flags:=mainwindowgads[ThirdGadget]^.Flags and ~GFLG_Selected;
                          dummy:=AddGList(mainwindow,mainwindowglist,dummy,~0,nil);
                          RefreshGList(mainwindowglist,mainwindow,nil,~0);
                          gt_setsinglegadgetattr(mainwindowgads[DisplayGadget],mainwindow,
                                                 GTTX_Text,long(@SecondGad[1]));
                        end;
                      ThirdGadget :
                        begin
                          dummy:=RemoveGList(mainwindow,mainwindowglist,~0);
                          mainwindowgads[FirstGadget]^.Flags:=mainwindowgads[FirstGadget]^.Flags and ~GFLG_Selected;
                          mainwindowgads[SecondGadget]^.Flags:=mainwindowgads[SecondGadget]^.Flags and ~GFLG_Selected;
                          mainwindowgads[ThirdGadget]^.Flags:=mainwindowgads[ThirdGadget]^.Flags or GFLG_Selected;
                          dummy:=AddGList(mainwindow,mainwindowglist,dummy,~0,nil);
                          RefreshGList(mainwindowglist,mainwindow,nil,~0);
                          gt_setsinglegadgetattr(mainwindowgads[DisplayGadget],mainwindow,
                                                 GTTX_Text,long(@ThirdGad[1]));
                        end;
                     end;
                    pimsg:=gt_getimsg(mainwindow^.userport);
                  end;
              until done;
              closewindowmainwindow;
            end
           else
            writeln('Could not open window.');
          Freeimages;
        end
       else
        writeln('Unable to make images.');
      closelibs;
    end
   else
    writeln('Could not open libraries.');
end.

File: gui/DesignerV1_55/ToggleDemo/ToggleDemo.c
/* Simple demo using a Designer created file */
/* Only run from CLI */

#include "toggledemowin.c"

ULONG MyTags[] =
{
	GTTX_Text,0,
	TAG_DONE
};

int main(void)
{
int done=0;
ULONG class;
UWORD code;
struct Gadget *pgsel;
struct IntuiMessage *imsg;
UWORD GadNumber;
if (OpenLibs()==0)
	{
	if (MakeImages()==0)
		{
		if (OpenWindowMainWindow()==0)
			{
			while (done==0)
				{
				Wait(1L << MainWindow->UserPort->mp_SigBit);
				imsg=GT_GetIMsg(MainWindow->UserPort);
				while (imsg != NULL )
					{
					class=imsg->Class;
					code=imsg->Code;
					pgsel=(struct Gadget *)imsg->IAddress; /* Only reference if it is a gadget message */
					GT_ReplyIMsg(imsg);
					GadNumber=99;
					switch (class)
						{
						case IDCMP_CLOSEWINDOW :
							done=1;
							break;
						case IDCMP_GADGETDOWN :
							GadNumber=pgsel->GadgetID;
							break;
						case IDCMP_VANILLAKEY :
							switch (code)
								{
								case 'F' :
									GadNumber = FirstGadget;
									break;
								case 'S' :
									GadNumber = SecondGadget;
									break;
								case 'T' :
									GadNumber = ThirdGadget;
									break;
								}
							break;
						}
					switch (GadNumber)
						{
						case FirstGadget :
							/* Remove gadgets from window */
							RemoveGList(MainWindow,MainWindowGList,~0);
							/* Change Gadget Flags */
							MainWindowGadgets[FirstGadget ]->Flags |=  GFLG_SELECTED;
						 	MainWindowGadgets[SecondGadget]->Flags &= ~GFLG_SELECTED;
							MainWindowGadgets[ThirdGadget ]->Flags &= ~GFLG_SELECTED;
							/* Put Gadgets Back in window */
							AddGList(MainWindow,MainWindowGList,50,~0,NULL);
							/* Refresh Gadgets */
							RefreshGList(MainWindowGList,MainWindow,NULL,~0);
							MyTags[1] = (ULONG)"First Option";
							GT_SetGadgetAttrsA(MainWindowGadgets[DisplayGadget],MainWindow, NULL, (struct TagItem *)MyTags);
							break;
						case SecondGadget :
							RemoveGList(MainWindow,MainWindowGList,~0);
							MainWindowGadgets[FirstGadget ]->Flags &= ~GFLG_SELECTED;
						 	MainWindowGadgets[SecondGadget]->Flags |=  GFLG_SELECTED;
							MainWindowGadgets[ThirdGadget ]->Flags &= ~GFLG_SELECTED;
							AddGList(MainWindow,MainWindowGList,0,~0,NULL);
							RefreshGList(MainWindowGList,MainWindow,NULL,~0);
							MyTags[1] = (ULONG)"Second Option";
							GT_SetGadgetAttrsA(MainWindowGadgets[DisplayGadget],MainWindow, NULL, (struct TagItem *)MyTags);
							break;
						case ThirdGadget :
							RemoveGList(MainWindow,MainWindowGList,~0);
							MainWindowGadgets[FirstGadget ]->Flags &= ~GFLG_SELECTED;
						 	MainWindowGadgets[SecondGadget]->Flags &= ~GFLG_SELECTED;
							MainWindowGadgets[ThirdGadget ]->Flags |=  GFLG_SELECTED;
							AddGList(MainWindow,MainWindowGList,50,~0,NULL);
							RefreshGList(MainWindowGList,MainWindow,NULL,~0);
							MyTags[1] = (ULONG)"Third Option";
							GT_SetGadgetAttrsA(MainWindowGadgets[DisplayGadget],MainWindow, NULL, (struct TagItem *)MyTags);
							break;
						}
					imsg=GT_GetIMsg(MainWindow->UserPort);
					}
				}
			CloseWindowMainWindow();
			}
		else
			printf("Cannot open window.\n");
		FreeImages();
		}
	else
		printf("Cannot make images.\n");
	}
else
	printf("Cannot open libraries.\n");
}
